backend/cfg/cfg_ls.ml

```diff
@@ -0,0 +1,280 @@
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+open! Cfg_regalloc_utils
+open! Cfg_ls_utils
+module State = Cfg_ls_state
+
+let snapshot_for_fatal = ref None
+
+module Utils = struct
+  include Cfg_ls_utils
+
+  let debug = ls_debug
+
+  let invariants = ls_invariants
+
+  let log = log
+
+  let log_body_and_terminator = log_body_and_terminator
+
+  let is_spilled reg = reg.Reg.spill
+
+  let set_spilled _reg = ()
+end
+
+let rewrite : State.t -> Cfg_with_liveness.t -> Reg.t list -> unit =
+ fun state cfg_with_liveness spilled_nodes ->
+  Cfg_regalloc_rewrite.rewrite_gen
+    (module State)
+    (module Utils)
+    state cfg_with_liveness spilled_nodes
+  |> ignore
+
+(* Equivalent to [build_intervals] in "backend/interval.ml". *)
+let build_intervals : State.t -> Cfg_with_liveness.t -> unit =
+ fun state cfg_with_liveness ->
+  log ~indent:1 "build_intervals";
+  let cfg_with_layout = Cfg_with_liveness.cfg_with_layout cfg_with_liveness in
+  let liveness = Cfg_with_liveness.liveness cfg_with_liveness in
+  let past_ranges : Interval.t Reg.Tbl.t = Reg.Tbl.create 123 in
+  let current_ranges : Range.t Reg.Tbl.t = Reg.Tbl.create 123 in
+  let add_range (reg : Reg.t) ({ begin_; end_ } as range : Range.t) : unit =
+    match Reg.Tbl.find_opt past_ranges reg with
+    | None ->
+      Reg.Tbl.replace past_ranges reg
+        { Interval.reg; begin_; end_; ranges = [range] }
+    | Some (interval : Interval.t) ->
+      interval.ranges <- range :: interval.ranges;
+      interval.end_ <- end_
+  in
+  let update_range (reg : Reg.t) ~(begin_ : int) ~(end_ : int) : unit =
+    match Reg.Tbl.find_opt current_ranges reg with
+    | None -> Reg.Tbl.replace current_ranges reg { Range.begin_; end_ }
+    | Some ({ begin_ = _; end_ = curr_end } as curr) ->
+      if (begin_ asr 1) - (curr_end asr 1) <= 1
+      then curr.end_ <- end_
+      else (
+        add_range reg curr;
+        Reg.Tbl.replace current_ranges reg { Range.begin_; end_ })
+  in
+  (* Equivalent to [update_interval_position_by_instr] in
+     "backend/interval.ml". *)
+  let update_instr :
+      type a.
+      int ->
+      a Cfg.instruction ->
+      trap_handler:bool ->
+      destroyed:Reg.t array ->
+      unit =
+   fun pos instr ~trap_handler ~destroyed ->
+    let on = pos lsl 1 in
+    let off = on + 1 in
+    if trap_handler
+    then
+      Array.iter Proc.destroyed_at_raise ~f:(fun reg ->
+          update_range reg ~begin_:on ~end_:on);
```

Comment:
```suggestion
          update_range reg ~begin_:off ~end_:off);
```

Reply:
I think upstream registers the destruction
on the `off` position of the `Iend` instruction
closing the protected block, while this CFG
version registers the destruction on the `on`
position of the first instruction of the handler.
I think it is OK to have the destruction on the
first instruction of the handler, and on the
CFG there is no direct equivalent to the
"previous" instruction because the layout
may change.